File: src/main/java/workshop/part3/futures/VettingFutureActor.java
package workshop.part3.futures;

import java.util.concurrent.CompletionStage;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.pattern.Patterns;
import akka.util.Timeout;
import scala.compat.java8.FutureConverters;
import scala.concurrent.duration.FiniteDuration;

import static workshop.part1.Verdict.VerdictType;

public class VettingFutureActor extends AbstractActor {

    private final ActorRef userActor;
    private final ActorRef fraudWordActor;
    private final FiniteDuration timeoutVetting;

    VettingFutureActor(ActorRef userActor, ActorRef fraudWordActor, FiniteDuration timeoutVetting) {
        this.userActor = userActor;
        this.fraudWordActor = fraudWordActor;
        this.timeoutVetting = timeoutVetting;
    }

    @Override
    // receive an Ad and combine the Futures returned from underlying actors to reply with a VerdictType
    public Receive createReceive() {
        return receiveBuilder()
            .build();
    }

    // utility for passing a VerdictType wrapped in javas CompletionStage to a receiver
    private void pipeTo(ActorRef receiver, CompletionStage<VerdictType> verdict) {
        Patterns.pipe(FutureConverters.toScala(verdict), context().system().dispatcher()).to(receiver);
    }

    // handy converter from scala Future to javas CompletionStage
    private <T> CompletionStage<T> ask(ActorRef receiver, Object msg) {
        return (CompletionStage<T>) FutureConverters.toJava(Patterns.ask(receiver, msg, new Timeout(timeoutVetting)));
    }
}

File: src/main/kotlin/workshop/part3/futures/VettingFutureActor.kt
package workshop.part3.futures

import akka.actor.AbstractActor
import akka.actor.ActorRef
import akka.pattern.Patterns
import akka.util.Timeout
import scala.compat.java8.FutureConverters
import scala.concurrent.duration.FiniteDuration
import workshop.common.ad.Ad
import workshop.common.userservice.UserCriminalRecord
import workshop.part1.VerdictType
import workshop.part2.FraudWordActor
import workshop.part2.UserActor.CheckUser
import workshop.part2.UserActor.CheckUserResult
import java.util.concurrent.CompletionStage

class VettingFutureActor internal constructor(private val userActor: ActorRef, private val fraudWordActor: ActorRef, private val timeoutVetting: FiniteDuration) : AbstractActor() {

    override fun createReceive(): AbstractActor.Receive {
        return receiveBuilder()
                .match(Ad::class.java) { ad ->
                    val userFuture = ask<CheckUserResult>(userActor, CheckUser(ad.userId))
                    val fraudWordFuture = ask<FraudWordActor.ExamineWordsResult>(fraudWordActor, FraudWordActor.ExamineWords(ad.toAdWords()))

                    val userOk = userFuture.thenApply { result -> result.record === UserCriminalRecord.GOOD }
                    val fraudWordOk = fraudWordFuture.thenApply { result -> result.fraudWords.isEmpty() }

                    val verdict = userOk.thenCombine(fraudWordOk) { user, fraudWord ->
                            listOf(user, fraudWord) }
                        .thenApply<VerdictType> { res -> if (res.all { r -> r }) VerdictType.GOOD else VerdictType.BAD }
                        .handle<VerdictType> { ok, throwable -> ok ?: VerdictType.PENDING}

                    pipeTo(sender(), verdict)
                }
                .build()
    }

    private fun pipeTo(receiver: ActorRef, verdict: CompletionStage<VerdictType>) {
        Patterns.pipe(FutureConverters.toScala(verdict), context().system().dispatcher()).to(receiver)
    }

    private fun <T> ask(receiver: ActorRef, msg: Any): CompletionStage<T> {
        return FutureConverters.toJava(Patterns.ask(receiver, msg, Timeout(timeoutVetting))) as CompletionStage<T>
    }
}
